lib/drivers/tandemDriver.js

```diff
@@ -838,72 +859,78 @@ module.exports = function (config) {
+                debug('recovering ', receive_seq, '(received ',pkt.payload['header_log_seq_no'], ')');
                 send_seq = receive_seq + 1;
-                debug('recovering ', receive_seq);
-                tandemCommand(COMMANDS.LOG_ENTRY_SEQ_REQ, [receive_seq], function (err) {
-                  if(err) {
-                    callback(err,null);
+              }
+
+              tandemCommand(COMMANDS.LOG_ENTRY_SEQ_REQ, [receive_seq], function (err) {
+                if(err) {
+                  callback(err,null);
+                }
+                retryRecoverTimer = setTimeout(function() {
+                  if(recovering) {
+                    debug('Retrying to recover..',receive_seq);
+                    tandemCommand(COMMANDS.LOG_ENTRY_SEQ_REQ, [receive_seq], function (err) {
+                      if(err) {
+                        callback(err,null);
+                      }
+                    });
                   }
-                });
-              } // drop out-of-order packets on the floor; they will be re-requested.
+                },1000);
+              });
             }
             else {
               if (recovering) {
                 debug('recovered ', receive_seq, pkt);
+                clearTimeout(retryRecoverTimer);
               }
               receive_seq = pkt.payload['header_log_seq_no'] + 1;
               recovering = false;
 
-              if(start !== end) {
-                // only update progress if not binary search
-                percentage = ((receive_seq-start)/(end-start) * 90)+10;
-                if(percentage > (prevPercentage+1)) {
-                  // only update progress to UI if there's an increase of at least 1 percent
-                  prevPercentage = percentage;
-                  progress(percentage);
-                }
+              percentage = ((receive_seq-start)/(end-start) * 90)+10;
+              if(percentage > (prevPercentage+1)) {
+                // only update progress to UI if there's an increase of at least 1 percent
+                prevPercentage = percentage;
+                progress(percentage);
               }
 
               if (receive_seq % 1000 === 0) {
                 debug('received ', receive_seq, ' of ', end);
               }
-              callback(null, pkt);
               if (receive_seq > end) {
                 if (__DEBUG__) {
                   var end_exec = Date.now();
                   var time = end_exec - start_exec;
                   debug('Execution time of tandemLogRequester: ' + time);
                 }
+                cfg.deviceComms.flush(); // making sure we flush the buffers
+                clearInterval(sendTimer);
                 clearInterval(listenTimer);
               }
+              callback(null, pkt);
             }
           }
         };
         processPacket(cfg.deviceComms.nextPacket());
-        delay.forEach(processPacket);
       }
-    }, 1);
+    }, 5);
```

Comment:
We have a `sendTimer` interval timer and a `listenTimer` interval timer that are running at the same time at a set frequency. I'm now renaming this frequency as `INTERVAL_FREQ` and applying it to the relevant timers to improve readability.